quickwit-actors/src/progress.rs

```diff
@@ -0,0 +1,183 @@
+/// It is used in lieu of healthcheck.
+///
+/// If no progress is observed until the next heartbeat, the actor will be killed.
+#[derive(Clone)]
+pub(crate) struct Progress(Arc<AtomicU32>);
+
+#[derive(Clone, Debug, Copy, PartialEq, Eq)]
+enum ProgressState {
+    // No update recorded since the last call to .check_for_update()
+    NoUpdate,
+    // An update was recorded since the last call to .check_for_update()
+    Updated,
+    // The actor is in the protected zone.
+    //
+    // The protected zone should seldom be used. It is useful
+    // when calling an external library that is blocking for instance.
+    //
+    // Another use case is blocking when sending a message to another actor
+    // with a saturated message bus.
+    // The failure detection is then considered to be the problem of
+    // the downstream actor.
+    //
+    // As long as the actor is in the protected zone, healthchecking won't apply
+    // to it.
+    //
+    // The value inside starts at 0.
+    ProtectedZone(u32),
+}
+
+impl Into<u32> for ProgressState {
+    fn into(self) -> u32 {
+        match self {
+            ProgressState::NoUpdate => 0,
+            ProgressState::Updated => 1,
+            ProgressState::ProtectedZone(level) => 2 + level,
+        }
+    }
+}
+
+impl From<u32> for ProgressState {
+    fn from(level: u32) -> Self {
+        match level {
+            0 => ProgressState::NoUpdate,
+            1 => ProgressState::Updated,
+            level => ProgressState::ProtectedZone(level - 2),
+        }
+    }
+}
+
+impl Default for Progress {
+    fn default() -> Progress {
+        Progress(Arc::new(AtomicU32::new(ProgressState::Updated.into())))
+    }
+}
+
+impl Progress {
+    pub fn record_progress(&self) {
+        self.0
+            .fetch_max(ProgressState::Updated.into(), Ordering::Relaxed);
+    }
+
+    pub fn protect_zone(&self) -> ProtectZoneGuard {
+        loop {
+            let previous_state: ProgressState = self.0.load(Ordering::SeqCst).into();
+            let new_state = match previous_state {
+                ProgressState::NoUpdate | ProgressState::Updated => ProgressState::ProtectedZone(0),
+                ProgressState::ProtectedZone(level) => ProgressState::ProtectedZone(level + 1),
+            };
+            if self
+                .0
+                .compare_exchange(
+                    previous_state.into(),
+                    new_state.into(),
+                    Ordering::SeqCst,
+                    Ordering::SeqCst,
+                )
+                .is_ok()
+            {
+                return ProtectZoneGuard(self.0.clone());
+            }
+        }
+    }
+
+    /// This method mutates the state as follows and returns true if
+    /// the object was in the protected zone or had change registered.
+    /// - Updated -> NoUpdate, returns true
+    /// - NoUpdate -> Updated, returns true
+    /// - ProtectedZone -> ProtectedZone, returns true
+    pub fn harvest_changes(&self) -> bool {
```

Comment:
had/registered_activity_since_last_call or similar would be a better name I think

Reply:
agreed